An example for distributing prebuilt binaries

hofstee opened the pull request:
I wouldn't merge this as is, but here's an example of using GitHub Actions CI to build `sv2v` binaries for Linux. You can see the end results on my fork here: https://github.com/hofstee/sv2v/releases

The binaries at least seem to work on my laptop, I'm not too familiar with Haskell and if it statically links things while building its binaries.

zachjs commented:
Thank you for working on this! I've been very interested in switching over to Github Actions, and figured it would be possible to use it for making the release binaries.

1. Is there a way this could be automatically triggered when a tag is created, rather than on every push? I figure I'll only want to make "official" releases occasionally.
2. Can it be set up to only make the release if tests pass?
3. Can it be made to build on multiple architectures and upload separate artifacts on the same release? It might be nice to release a Windows binary too.

My final concern is in respecting the licenses of the Haskell libraries in use. I believe some of them may require that their license be included in the resulting binary. I can look into this separately.

I believe all of the Haskell parts are statically linked on the resulting binary.

hofstee commented:
I think this should do it, or at least be fairly close to something you can work with.

It expects tags to be `v*`, and that's what triggers the CI builds. I see you've migrated your other CI over to GitHub Actions so I think you're probably familiar enough with how things work now.

The documentation says that if the `if:` check on each step is blank, it defaults to `success()`, so a step shouldn't run if the previous steps have failed. With that in mind I've structured this in the following manner:

- The build and testing happens the same way it happens on your pipeline.
- After make test completes, it uploads an artifact for each os in the build matrix.
- A second job depends on the build job, and runs on linux. This downloads all the artifacts, creates a release, then uploads the release artifacts.

I've structured it like this because we need to only create the release once, or else it will run into problems where that step fails. If you want some other ideas there's [an issue on the create-release action](https://github.com/actions/create-release/issues/14) that has some approaches others took.

The main thing I'm not a huge fan of is that there is a step for every OS in the build matrix for downloading artifacts and uploading the release artifacts. This was the only way I could think of to only trigger when all platforms build successfully. I'm not sure what the job status is of build if ubuntu succeeds and macos fails, but maybe if that sets it to fail then you could define two jobs, where the first creates the release, and the second runs on the build matrix again and uploads the release artifacts.

zachjs commented:
This looks like great progress! I have a few final requests:

* I think we can lump this in with the existing `test.yaml` workflow (which has changed a little) and have the release creation job conditioned on the ref/tag name. The artifacts can be created unconditionally.
* I've added a `NOTICE` file containing the licenses of sv2v's dependencies. I think the release binaries should be bundled with `NOTICE`, `LICENSE`, and `README.md`, ideally inside a folder. 
* Can you standardize the artifact names around simply `macos` and `linux`?
* Can the commits be rebased and squashed?

Thank you for all the work you've put in to this!

hofstee commented:
> I think we can lump this in with the existing test.yaml workflow (which has changed a little) and have the release creation job conditioned on the ref/tag name.

I'm struggling to figure out how to change `on: push: tags: ...` to a `job.<jobname>.if`. The `test.yaml` workflow already only triggers on pushes so that's ok, but I'm not sure how to get the tag here just for the release job.

> I think the release binaries should be bundled with NOTICE, LICENSE, and README.md, ideally inside a folder.

I set it up to copy the files to `bin` before packaging up the artifact. For the folder name, I set it to `sv2v-<os>` now since that seemed like the most straightforward thing without having to copy things around to other directories or ending up with a random folder called `Linux` once you unzip.

> Can you standardize the artifact names around simply macos and linux?

I've changed it to use `runner.os`, so the resulting artifacts are named `Linux` and `macOS` everywhere now.

> Can the commits be rebased and squashed?

It might be easiest to do that if you merge the PR through the GitHub page, usually that's one of the options in the dropdown menu. If not I can do so once everything looks good.

zachjs commented:
I've made some changes to your example to get it ready. The end result is at https://github.com/zachjs/sv2v/tree/release.

* Rather than having the workflow create the release, the workflow is triggered on release creation. I use `if: github.event_name == 'release'` to skip the artifact ion creation steps and the release job.
* The artifacts lose their permissions. I added a `chmod +x` to make the executables executable again.

I'd say this is ready to go. Thank you for all the work you've put into this! I'd like for you to get credit for your contribution. You can update this PR to use the finalized diff, or I can update the author on that `release` commit, whichever you'd prefer.

hofstee commented:
Updating the author on that commit is probably the simpler of the two. Thanks for finishing it up!

zachjs commented:
This has been merged as https://github.com/zachjs/sv2v/commit/bbd072d41d8aba9bf13f5680b3f4b6cd6fe2b379. Thanks again for your contribution!
